cli/src/cmd/watch.rs

```diff
@@ -0,0 +1,362 @@
+    event::{Event, ProcessEnd},
+    handler::SyncFnHandler,
+    paths::summarise_events_to_env,
+    signal::source::MainSignal,
+    Watchexec,
+};
+
+use crate::utils;
+
+/// Executes a [`Watchexec`] that listens for changes in the project's src dir and reruns `forge
+/// build`
+pub async fn watch_build(args: BuildArgs) -> eyre::Result<()> {
+    let (init, mut runtime) = args.watchexec_config()?;
+    let cmd = cmd_args(args.watch.watch.as_ref().map(|paths| paths.len()).unwrap_or_default());
+    runtime.command(cmd.clone());
+
+    let wx = Watchexec::new(init, runtime.clone())?;
+    on_action(args.watch, runtime, Arc::clone(&wx), cmd, (), |_| {});
+
+    // start executing the command immediately
+    wx.send_event(Event::default()).await?;
+    wx.main().await??;
+
+    Ok(())
+}
+
+/// Executes a [`Watchexec`] that listens for changes in the project's src dir and reruns `forge
+/// test`
+pub async fn watch_test(args: TestArgs) -> eyre::Result<()> {
+    let (init, mut runtime) = args.build_args().watchexec_config()?;
+    let cmd = cmd_args(
+        args.build_args().watch.watch.as_ref().map(|paths| paths.len()).unwrap_or_default(),
+    );
+    runtime.command(cmd.clone());
+    let wx = Watchexec::new(init, runtime.clone())?;
+
+    // marker to check whether we can safely override the command
+    let has_conflicting_pattern_args =
+        args.filter().pattern.is_some() || args.filter().test_pattern.is_some();
+    on_action(
+        args.build_args().watch.clone(),
+        runtime,
+        Arc::clone(&wx),
+        cmd,
+        has_conflicting_pattern_args,
+        on_test,
+    );
+
+    // start executing the command immediately
+    wx.send_event(Event::default()).await?;
+    wx.main().await??;
+
+    Ok(())
+}
+
+/// The `on_action` hook for `forge test --watch`
+fn on_test(action: OnActionState<bool>) {
+    let OnActionState { args, runtime, action, wx, cmd, other } = action;
+    let has_conflicting_pattern_args = other;
+    if has_conflicting_pattern_args {
+        // can't set conflicting arguments
+        return
+    }
+
+    let mut cmd = cmd.clone();
+    // get changed files and update command accordingly
+    let sol_files: Vec<_> = action
+        .events
+        .iter()
+        .flat_map(|e| e.paths())
+        .filter(|(path, _)| path.extension() == Some(OsStr::new("sol")))
+        .filter_map(|(path, _)| path.to_str())
+        .collect();
+
+    if sol_files.is_empty() {
+        return
+    }
+
+    // replace `--match-path` | `-mp` argument
+    if let Some(pos) = cmd.iter().position(|arg| arg == "--match-path" || arg == "-mp") {
+        // --match-path requires 1 argument
+        cmd.drain(pos..=(pos + 1));
+    }
+
+    // append `--match-path` regex
+    let re_str = format!("({})", sol_files.join("|"));
+    if let Ok(re) = Regex::from_str(&re_str) {
+        let mut new_cmd = cmd.clone();
+        new_cmd.push("--match-path".to_string());
+        new_cmd.push(re.to_string());
```

Comment:
ok so even if the user has specified a match-path argument, it gets overwritten with the files that were changed in this run. i guess there's an edge case here, where if I have supplied `--match-path A.sol` and the file that changed is B.sol, tests for B.sol end up getting run, when I only wanted A.sol? same issue with inheritance etc., seems tricky to cover extensively, maybe we should document it as a known limitation?

Reply:
yeah that could be unexpected, so I assume the correct way would be to leave the command unchanged if it was initially run with a --match-path.